### PR TITLE
Use modern setuptools with static setup.cfg

### PR DESCRIPTION
**Pull request by aglavic**
Dependencies are now properly included when installing with pip install.
Move the .deb generation into a separate bash script (from setup.py).

### CODE REVIEW

File: genx/make_deb.sh [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#!/bin/sh
```
Comment (aglavic) [inline genx/make_deb.sh:1]: As already separated from setup script please move to debian_build folder.

File: genx/make_deb.sh [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+cp ../../debian_build/postrm debian/
+
+# create .deb package
+
+PY_VERSION=$(python -c "import sys; print(sys.version_info.major + sys.version_info.minor)")
+
+dpkg-buildpackage -i.* -I -rfakeroot -us -uc > ../last_package.log
+cd ..
+
+#os.rename((__name__+'_'+__version__).lower()+'-1_all.deb', __name__+'-'+__version__+'_py%i%i.deb'%(
```
Comment (aglavic) [inline genx/make_deb.sh:58]: The script should be package specific, i.e. include the current version number. The renaming is also important for distribution and should be part of this script. (A reason this was in the setup.py script.)

File: genx/setup.cfg [hunk 3]
```diff
@@ -1,14 +1,51 @@
...
+    genx.lib
+    genx.models
+    genx.models.lib
+    genx.plugins
+    genx.plugins.add_ons
+    genx.plugins.add_ons.help_modules
+    genx.plugins.data_loaders
+    genx.plugins.data_loaders.help_modules
+scripts = scripts/genx
+install_requires =
```
Comment (aglavic) [inline genx/setup.cfg:32]: There used to be a distiction between "requires" and "install_requires". The packages you list are required to run, not install genx.

File: genx/setup.cfg [hunk 4]
```diff
@@ -1,14 +1,51 @@
...
+    numpy
+    matplotlib
+    scipy
+    appdirs
+    wxPython
```
Comment (aglavic) [inline genx/setup.cfg:37]: You should remove the wxPython requirement. It won't work with pip anyway and with the development version there is a scripting API that doesn't require wx at all.

File: genx/setup.cfg [hunk 5]
```diff
@@ -1,14 +1,51 @@
...
+    h5py
```
Comment (aglavic) [inline genx/setup.cfg:39]: I think there are a few other requirements (numba, psutils), some of which for additional functionality (pymysql, vtk), is it possible to enter recommended packages?

File: genx/setup.cfg [hunk 6]
```diff
@@ -1,14 +1,51 @@
-[sdist]
-formats=gztar
+[metadata]
+name = GenX
+version = attr: genx.version.__version__
+author = Matts Bjorck, Artur Glavic
+author_email = artur.glavic@gmail.com
```
Comment (aglavic) [inline genx/setup.cfg:5]: artur.glavic@psi.ch

File: genx/setup.cfg [hunk 7]
```diff
@@ -1,14 +1,51 @@
...
+license_file = LICENSE.txt
+classifiers =
+    Programming Language :: Python :: 3
+    License :: OSI Approved :: GNU General Public License v3 (GPLv3)
+    Operating System :: OS Independent
+    Topic :: Scientific/Engineering :: Physics
+    Development Status :: 6 - Mature
 
-[bdist]
-formats=rpm
```
Comment (aglavic) [inline genx/setup.cfg:5]: Is there any reason not to use rpm as default bdist? Mac and Windows will use pyinstaller, anyway.

File: genx/setup.cfg [hunk 8]
```diff
@@ -1,14 +1,51 @@
-[sdist]
-formats=gztar
```
Comment (aglavic) [inline genx/setup.cfg:2]: This is to use .tar.gz as default sdist, why did you delete it?

File: genx/setup.cfg [hunk 9]
```diff
@@ -1,14 +1,51 @@
...
+    matplotlib
+    scipy
+    appdirs
+    wxPython
+    h5py
 
-[bdist_wininst]
 
-[install]
-optimize=1
```
Comment (aglavic) [inline genx/setup.cfg:14]: This used to be required to build binary compiled code for install, is that not necessary anymore?

File: genx/setup.cfg [hunk 10]
```diff
@@ -1,14 +1,51 @@
...
+[options.package_data]
+genx =
+    genx.conf
+    examples/*.*
+    LICENSE.txt
+    changelog.txt
```
Comment (aglavic) [inline genx/setup.cfg:46]: changelog.txt and LICENSE.txt should be in the root folder, not the genx package. (genx here is relative to setup.py, thus it corresponds to genx/genx/ in the repository.)
As you have created a license file already in the base folder I think it makes sense to put all of that in the base folder to separate from the package or all of it in the package.

File: genx/make_deb.sh [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+set -e
+
+python setup.py bdist_rpm
+
+# These commands requires dpkg-dev alien
+echo "Moving distribution files..."
+
+cd dist
+
+NAME=$(basename *.tar.gz .tar.gz)
```
Comment (aglavic) [inline genx/make_deb.sh:13]: This will fail when ever there is more than one .tar.gz file in the dist folder.

**Review by aglavic — COMMENTED** [review]
I've made a few comments in the code, can you check those and make necessary changes before I merge it to the master branch?

File: genx/setup.cfg (hunk 3 above)
Comment (llohse) [reply genx/setup.cfg:32]: According to https://packaging.python.org/guides/distributing-packages-using-setuptools/#install-requires, install_requires specifies "what dependencies a project minimally needs to run" (as opposed to build_requires).

File: genx/setup.cfg (hunk 4 above)
Comment (llohse) [reply genx/setup.cfg:37]: Good point. Will do.

File: genx/setup.cfg (hunk 5 above)
Comment (llohse) [reply genx/setup.cfg:39]: None, that I am aware of unfortunately.

File: genx/setup.cfg (hunk 7 above)
Comment (llohse) [reply genx/setup.cfg:5]: It just felt counterintuitive to me, as you can always use setup.py bdist_rpm to generate the rpm (as in make_deb.sh). If you liked it, I can revert it.

File: genx/make_deb.sh (hunk 1 above)
Comment (llohse) [reply genx/make_deb.sh:1]: ok

File: genx/make_deb.sh (hunk 2 above)
Comment (llohse) [reply genx/make_deb.sh:58]: ok

File: genx/setup.cfg (hunk 8 above)
Comment (llohse) [reply genx/setup.cfg:2]: It is the default on linux anyway [1]. On windows it defaults to zip.

[1] https://docs.python.org/3/distutils/sourcedist.html

File: genx/setup.cfg (hunk 9 above)
Comment (llohse) [reply genx/setup.cfg:14]: I have not seen this anywhere before and I could not find any documentation on this option.
